### ISSUE TITLE
Ansible 2.9 networking can be slower than Ansible 2.2.

### ISSUE THREAD

**Issue by kjoyce77** [ansible/ansible#70068]
##### SUMMARY

Task executor subprocesses to update an active persistent connection.  When using more than about 5 forks per core this causes current Ansible versions to actually run slower than Ansible 2.2 using the local connection plugin.  @gundalow suggested I start with an issue here from my mention of this general performance issue on reddit.com.  I can prepare a PR if it is decided this is an issue that needs fixing.

I've found considerable performance gains from monkey patching the start_connection function from ansible.executor.task_executor.  This is only a copy and paste from ansbile.executor.task_executor and ansible.cli.scripts.ansible_connection_cli_stub.  The only code I added was a try/except that tries to connect to the existing socket and update it before calling the original function to subprocess a new one.

I've attached the monkey patch I used [here](https://github.com/ansible/ansible/files/4779818/linear.zip).

##### ISSUE TYPE
- Bug Report

##### COMPONENT NAME

bin/ansible-connection
lib/ansible/cli/scripts/ansible_connection_cli_stub.py
lib/ansible/executor/task_executor.py
Plugin connection/network_cli.py

##### ANSIBLE VERSION

```paste below
ansible 2.9.6
  config file = None
  configured module search path = ['/Users/f6oikxs/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/f6oikxs/playbooks/venv/lib/python3.6/site-packages/ansible
  executable location = /Users/f6oikxs/playbooks/venv/bin/ansible
  python version = 3.6.6 (default, Aug 20 2018, 15:09:41) [GCC 4.2.1 Compatible Apple LLVM 9.1.0 (clang-902.0.39.2)]
```

##### CONFIGURATION

```paste below
ANSIBLE_PIPELINING(/home/f6oikxs/kevin_playbooks/show_clock/ansible.cfg) = True
DEFAULT_CALLBACK_WHITELIST(/home/f6oikxs/kevin_playbooks/show_clock/ansible.cfg) = ['profile_tasks']
DEFAULT_JINJA2_NATIVE(/home/f6oikxs/kevin_playbooks/show_clock/ansible.cfg) = True
HOST_KEY_CHECKING(env: ANSIBLE_HOST_KEY_CHECKING) = False
INTERPRETER_PYTHON(/home/f6oikxs/kevin_playbooks/show_clock/ansible.cfg) = /usr/bin/env python
PERSISTENT_CONNECT_TIMEOUT(/home/f6oikxs/kevin_playbooks/show_clock/ansible.cfg) = 120
RETRY_FILES_ENABLED(/home/f6oikxs/kevin_playbooks/show_clock/ansible.cfg) = True

```

##### OS / ENVIRONMENT

All persistent network connections.

##### STEPS TO REPRODUCE

I use the below playbook and run it against an inventory file with 80 hosts.
ansible-playbook playbook.yml -i inventory/inventory.ini  --forks 40

```yaml
- name: RUN COMMANDS
  # 80 devices, running with 40 forks
  hosts: test_routers
  gather_facts: no
  connection: network_cli
  tasks:
    - cli_command:
        command: show clock
        
    - cli_command:
        command: show version
      register: result
      
    - debug:
        msg: "{{ result.stdout_lines[-1] }}"
```

##### EXPECTED RESULTS

I would expect it to run faster than Ansible 2.2.  Using ios_command in Ansible 2.2 here are the results I achieved.

Monday 15 June 2020  10:51:41 +0000 (0:00:02.465)       0:00:38.343 *********** 

ios_command ------------------------------------------------------------ 17.13s
ios_command ------------------------------------------------------------ 15.96s
debug ------------------------------------------------------------------- 2.72s
debug ------------------------------------------------------------------- 2.47s

Alternatively, with my monkey patch here are the results:

Monday 15 June 2020  10:23:56 +0000 (0:00:01.527)       0:00:47.595 *********** 

cli_command ------------------------------------------------------------ 35.97s
cli_command ------------------------------------------------------------- 8.33s
debug ------------------------------------------------------------------- 1.53s
debug ------------------------------------------------------------------- 1.49s

##### ACTUAL RESULTS

Instead using either network_cli or the local connection Ansible 2.9 runs slower.  Note how terrible the execution time is for a simple debug task.

Monday 15 June 2020  10:30:22 +0000 (0:00:21.253)       0:01:49.002 *********** 

cli_command ------------------------------------------------------------- 37.38s
cli_command ------------------------------------------------------------- 28.46s
debug ------------------------------------------------------------------- 21.64s
debug ------------------------------------------------------------------- 21.25s

```paste below
If requested I need to remove confidential data from any command output before posting here.
```

**Comment by ansibot** [comment]
Files identified in the description:
None

If these files are incorrect, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by sivel** [comment]
!component +bin/ansible-connection

**Comment by ansibot** [comment]
Files identified in the description:
* [`bin/ansible-connection`](https://github.com/ansible/ansible/blob/devel/bin/ansible-connection)

If these files are incorrect, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by kjoyce77** [comment]
!component lib/ansible/executor/task_executor.py

**Comment by sivel** [comment]
!component +lib/ansible/cli/scripts/ansible_connection_cli_stub.py

Trying to get the bot to classify this as support:networking, by adding a network component. Just an FYI.

**Comment by ansibot** [comment]
Files identified in the description:
* [`bin/ansible-connection`](https://github.com/ansible/ansible/blob/devel/bin/ansible-connection)
* [`lib/ansible/cli/scripts/ansible_connection_cli_stub.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/cli/scripts/ansible_connection_cli_stub.py)
* [`lib/ansible/executor/task_executor.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/executor/task_executor.py)

If these files are incorrect, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by ansibot** [comment]
Files identified in the description:
* [`bin/ansible-connection`](https://github.com/ansible/ansible/blob/devel/bin/ansible-connection)
* [`lib/ansible/cli/scripts/ansible_connection_cli_stub.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/cli/scripts/ansible_connection_cli_stub.py)
* [`lib/ansible/executor/task_executor.py`](https://github.com/ansible/ansible/blob/devel/lib/ansible/executor/task_executor.py)
* [`lib/ansible/plugins/connection`](https://github.com/ansible/ansible/blob/devel/lib/ansible/plugins/connection)

If these files are incorrect, please update the `component name` section of the description or use the `!component` bot command.

[click here for bot help](https://github.com/ansible/ansibullbot/blob/master/ISSUE_HELP.md)

**Comment by vladanmilosevic73** [comment]
We are in the same boat - any news on the options aside from what @kjoyce77 suggested. Any feedback?

**Comment by kjoyce77** [comment]
I've discovered if you specifically set the debug step to use the local connection method it resolves the slow non-network_cli steps.  Of course we still have the problem of multiple calls to network_cli actually being slowed down as soon as a cpu constraint is hit.

Also, I've discovered forking is actually copy-on-write.  So it saves memory and works a lot faster all around if the ansible-connection code is imported and the main process is forked as needed.  Running a brand-new instance of Python takes a huge performance hit vs. forking the existing one.

We had big consultants in who literally admitted Ansible does not work for some things with regards to network automation because of poor performance.  I am not the only one feeling the effects of this architecture.
